### PR TITLE
[Feature] Resume from the latest checkpoint automatically.

### PR DESCRIPTION
**Pull request by Czm369**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Resume from the latest checkpoint automatically  without specifying the path of the checkpoint.

## Modification

Add a new argument `auto-resume` in tools/train.py and add a new function  `find_latest_checkpoint` in mmdet/utils/misc.py.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/utils/misc.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import glob
+import os.path as osp
+
+
+def find_latest_checkpoint(path, ext='pth'):
```
Comment (hhaAndroid) [inline mmdet/utils/misc.py:6]: Please add unit test first

File: tools/train.py [hunk 2]
```diff
@@ -104,6 +108,7 @@ def main():
                                 osp.splitext(osp.basename(args.config))[0])
     if args.resume_from is not None:
         cfg.resume_from = args.resume_from
+    cfg.auto_resume = args.auto_resume
```
Comment (hhaAndroid) [inline tools/train.py:111]: if args.auto_resume is not None:
        cfg.auto_resume = args.auto_resume

File: mmdet/apis/train.py [hunk 3]
```diff
@@ -196,6 +196,12 @@ def train_detector(model,
         runner.register_hook(
             eval_hook(val_dataloader, **eval_cfg), priority='LOW')
 
+    resume_from = None
+    if cfg.resume_from is None and cfg.auto_resume:
```
Comment (hhaAndroid) [inline mmdet/apis/train.py:200]:
```python
resume_from=None
if cfg.resume_from is not None and cfg.get('auto_resume',False):
   warnings.warn(..)
   resume_from=cfg.resume_from
elif cfg.resume_from is not None:
   resume_from=cfg.resume_from
elif cfg.get('auto_resume',False) is True:
   resume_from=find_latest_checkpoint()

if resume_from:
  ...
```
...

File: mmdet/utils/misc.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    """
+    References:
+        .. [1] https://github.com/microsoft/SoftTeacher
+                  /blob/main/ssod/utils/patch.py
+    """
+    if not osp.exists(path):
```
Comment (hhaAndroid) [inline mmdet/utils/misc.py:12]: Can not simply return under abnormal conditions, but output relevant information.

File: mmdet/utils/misc.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import glob
+import os.path as osp
+
+
+def find_latest_checkpoint(path, ext='pth'):
+    """
```
Comment (hhaAndroid) [inline mmdet/utils/misc.py:7]: docstr

File: mmdet/utils/misc.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import glob
+import os.path as osp
+
+
```
Comment (hhaAndroid) [inline mmdet/utils/misc.py:5]: ext -> suffix

File: tests/test_utils/test_misc.py [hunk 7]
```diff
@@ -160,3 +163,35 @@ def test_filter_scores_and_topk():
...
         torch.tensor([[0.4, 0.7], [0.1, 0.1], [0.4, 0.7], [0.5, 0.1]]))
+
+
+def test_find_latest_checkpoint():
+    with tempfile.TemporaryDirectory() as tmpdir:
+        path = tmpdir + '/none'
+        latest = find_latest_checkpoint(path)
+        assert latest is None
+
+    with tempfile.TemporaryDirectory() as tmpdir:
```
Comment (hhaAndroid) [inline tests/test_utils/test_misc.py:174]: Maybe you can use `tempfile.TemporaryFile`

File: mmdet/apis/train.py (hunk 3 above)
Comment (RangiLyu) [inline mmdet/apis/train.py:200]: Do not use `cfg.auto_resume`, it will cause BC breaking. Use cfg.get('auto_resume') instead.

Comment (Czm369) [reply mmdet/apis/train.py]: Done.

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**